tests/lib/prepare-uc20.sh

```diff
@@ -0,0 +1,281 @@
+Documentation=man:snap(1)
+
+[Service]
+Type=oneshot
+ExecStart=/usr/lib/snapd/snapd.spread-tests-run-mode-tweaks.sh
+RemainAfterExit=true
+
+[Install]
+WantedBy=multi-user.target
+EOF
+
+cat > "$snapddir/usr/lib/snapd/snapd.spread-tests-run-mode-tweaks.sh" <<'EOF'
+#!/bin/sh
+set -e
+# Print to kmsg and console
+# $1: string to print
+print_system()
+{
+    printf "%s spread-tests-run-mode-tweaks.sh: %s\n" "$(date -Iseconds --utc)" "$1" |
+        tee -a /dev/kmsg /dev/console /run/mnt/ubuntu-seed/spread-tests-run-mode-tweaks-log.txt || true
+}
+
+# ensure we don't enable ssh in install mode or spread will get confused
+if ! grep 'snapd_recovery_mode=run' /proc/cmdline; then
+    print_system "not in run mode - script not running"
+    exit 0
+fi
+if [ -e /root/spread-setup-done ]; then
+    print_system "already ran, not running again"
+    exit 0
+fi
+
+print_system "in run mode, not run yet, extracting overlay data"
+
+# extract data from previous stage
+(cd / && tar xf /run/mnt/ubuntu-seed/run-mode-overlay-data.tar.gz)
+cp -r /root/test-var/lib/extrausers /var/lib
+
+# user db - it's complicated
+for f in group gshadow passwd shadow; do
+    # now bind mount read-only those passwd files on boot
+    cat >/etc/systemd/system/etc-"$f".mount <<EOF2
+[Unit]
+Description=Mount root/test-etc/$f over system etc/$f
+Before=ssh.service
+
+[Mount]
+What=/root/test-etc/$f
+Where=/etc/$f
+Type=none
+Options=bind,ro
+
+[Install]
+WantedBy=multi-user.target
+EOF2
+    systemctl enable etc-"$f".mount
+    systemctl start etc-"$f".mount
+done
+
+mkdir -p /home/test
+chown 12345:12345 /home/test
+mkdir -p /home/ubuntu
+chown 1000:1000 /home/ubuntu
+mkdir -p /etc/sudoers.d/
+echo 'test ALL=(ALL) NOPASSWD:ALL' >> /etc/sudoers.d/99-test-user
+echo 'ubuntu ALL=(ALL) NOPASSWD:ALL' >> /etc/sudoers.d/99-ubuntu-user
+# TODO: do we need this for our nested VM? We don't login as root to the nested
+#       VM...
+sed -i 's/\#\?\(PermitRootLogin\|PasswordAuthentication\)\>.*/\1 yes/' /etc/ssh/sshd_config
+echo "MaxAuthTries 120" >> /etc/ssh/sshd_config
+grep '^PermitRootLogin yes' /etc/ssh/sshd_config
+systemctl reload ssh
+
+print_system "done setting up ssh for spread test user"
+
+touch /root/spread-setup-done
+EOF
+chmod 0755 "$snapddir/usr/lib/snapd/snapd.spread-tests-run-mode-tweaks.sh"
+
+snap pack --filename=snapd.snap "$snapddir"
+rm -r $snapddir
+
+# next re-pack the kernel snap with this version of ubuntu-core-initramfs
+
+# extract the kernel snap, including extracting the initrd from the kernel.efi
+kerneldir=/tmp/kernel-workdir
+"$TESTSLIB/repack-kernel.sh" extract upstream-pc-kernel.snap $kerneldir
+
+# extract snap-bootstrap from the snapd snap and put into the kernel work dir
+cp /tmp/snap-bootstrap-from-snapd-snap $kerneldir/skeleton/main/usr/lib/snapd/snap-bootstrap
+
+# copy the skeleton from our installed ubuntu-core-initramfs into the initrd 
+# skeleton for the kernel snap
+cp -ar /usr/lib/ubuntu-core-initramfs/main/* "$kerneldir/skeleton/main"
```

Comment:
this in effect means that we are using the same version of kernel modules from the snap that were already there, the same kernel, a new snap-bootstrap from edge channel of snapd, and a new version of the ubuntu-core-initramfs layout from spread/the PR under test.

I seem to remember there being some thinking that what we were doing in snapd was not great and that we should be using some other location/source for the kernel modules or something, but perhaps I am just misremembering

Reply:
there was, and when we tried that, we found bugs =) which were not fixed yet.

It would be nicer to execute ubuntu-core-initramfs to create and sign kernel.efi in one go, for the kernel/modules extracted from the snap. But that's not possible yet.

Also not sure if during this test we want to use existing snap-bootstrap from inside the kernel-snap, the one from snapd edge, or the one that is "current" / stable / vendored inside ubuntu-core-initramfs.

Reply:
yes I also am not sure what the right thing to do is, you probably saw in the PR, but I actually changed it slightly after opening the PR to not stage snap-bootstrap from the snapd snap into the kernel snap's initramfs, but actually use the version of snap-bootstrap that comes from snapd from the snappy-dev/image PPA so that this test tests exactly what ubuntu-core-initramfs would do if it was built today.

that being said it would still be kinda nice if we have testing somewhere here like a nightly or something that ran with ubuntu-core-initramfs from main here, but also staged snap-bootstrap from edge snapd snap into it too to make sure that continues to work and doesn't regress as edge changes